dataRetriever/resolvers/genericBlockBodyResolver.go

```diff
@@ -141,89 +140,88 @@ func (gbbRes *GenericBlockBodyResolver) RequestDataFromHashArray(hashes [][]byte
+	marshalizedMiniBlocks, missingMiniBlocksHashes := gbbRes.getMiniBlocks(hashes)
+	miniBlocks := make(block.MiniBlockSlice, 0)
 
+	for hash, marshalizedMiniBlock := range marshalizedMiniBlocks {
+		miniBlock := &block.MiniBlock{}
+		err := gbbRes.marshalizer.Unmarshal(miniBlock, marshalizedMiniBlock)
 		if err != nil {
 			log.Debug(err.Error())
-			gbbRes.miniBlockPool.Remove(hashes[i])
-			err = gbbRes.miniBlockStorage.Remove(hashes[i])
+			gbbRes.miniBlockPool.Remove([]byte(hash))
+			err = gbbRes.miniBlockStorage.Remove([]byte(hash))
 			if err != nil {
 				log.Debug(err.Error())
 			}
 
-			return nil
+			missingMiniBlocksHashes = append(missingMiniBlocksHashes, []byte(hash))
+			continue
 		}
 
-		expandedMiniBlocks[i] = mb
+		miniBlocks = append(miniBlocks, miniBlock)
 	}
 
-	return expandedMiniBlocks
+	return miniBlocks, missingMiniBlocksHashes
 }
 
 // getMiniBlocks method returns a list of serialized mini blocks from a given hash list either from data pool or from storage
-func (gbbRes *GenericBlockBodyResolver) getMiniBlocks(hashes [][]byte) [][]byte {
-	miniBlocks := gbbRes.getMiniBlocksFromCache(hashes)
+func (gbbRes *GenericBlockBodyResolver) getMiniBlocks(hashes [][]byte) (map[string][]byte, [][]byte) {
+	marshalizedMiniBlocks, missingMiniBlocksHashes := gbbRes.getMiniBlocksFromCache(hashes)
+	if len(missingMiniBlocksHashes) == 0 {
+		return marshalizedMiniBlocks, missingMiniBlocksHashes
+	}
 
-	if miniBlocks != nil {
-		return miniBlocks
+	marshalizedMiniBlocksFromStorer, missingMiniBlocksHashes := gbbRes.getMiniBlocksFromStorer(missingMiniBlocksHashes)
+	for hash, marshalizedMiniBlockFromStorer := range marshalizedMiniBlocksFromStorer {
+		marshalizedMiniBlocks[hash] = marshalizedMiniBlockFromStorer
 	}
 
-	return gbbRes.getMiniBlocksFromStorer(hashes)
+	return marshalizedMiniBlocks, missingMiniBlocksHashes
 }
 
-// getMiniBlocksFromCache returns a full list of miniblocks from cache.
-// If any of the miniblocks is missing the function returns nil
-func (gbbRes *GenericBlockBodyResolver) getMiniBlocksFromCache(hashes [][]byte) [][]byte {
-	miniBlocksLen := len(hashes)
-	miniBlocks := make([][]byte, miniBlocksLen)
-
-	for i := 0; i < miniBlocksLen; i++ {
-		cachedMB, _ := gbbRes.miniBlockPool.Peek(hashes[i])
+// getMiniBlocksFromCache returns a list of marshalized mini blocks from cache and a list of missing hashes
+func (gbbRes *GenericBlockBodyResolver) getMiniBlocksFromCache(hashes [][]byte) (map[string][]byte, [][]byte) {
+	marshalizedMiniBlocks := make(map[string][]byte)
+	missingMiniBlocksHashes := make([][]byte, 0)
 
-		if cachedMB == nil {
-			return nil
+	for i := 0; i < len(hashes); i++ {
+		miniBlock, ok := gbbRes.miniBlockPool.Peek(hashes[i])
+		if !ok {
+			missingMiniBlocksHashes = append(missingMiniBlocksHashes, hashes[i])
+			continue
 		}
 
-		buff, err := gbbRes.marshalizer.Marshal(cachedMB)
+		buff, err := gbbRes.marshalizer.Marshal(miniBlock)
```

Comment:
no need to marshal, send direct miniblock back.

Reply:
Refactored